### PR TITLE
MARGARITA-CLI-HTTP_SERVER

### PR DESCRIPTION
**Pull request by RitaGlushkova**

### CODE REVIEW

File: cli-files/go-cat/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		Short: "go-cat command will output the contents of a file",
+		Long: `go-cat command that takes a path to a file as an argument, 
+	then opens that file and prints it out.
+ `,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) < 1 {
+				fmt.Println("Missing parameter, provide file name!")
+				return nil
+			}
+			path := args[0]
```
Comment (illicitonion) [inline cli-files/go-cat/cmd/root.go:26]: What happens if you pass more than one argument to `go-cat`? If someone gives you some information, it's best not to ignore it - typically you'd either do something useful with the other arguments, or you'd tell the user they made a mistake (by printing a message to stderr) and error (by exiting with a non-zero exit code).

In particular, we tend not to just print a warning and continue because a lot of times programs running in the terminal aren't running where someone's watching them, e.g. if you used `go-cat` in part of a GitHub Action to verify something about your code, if it printed a warning but continued and didn't signal failure with its exit code, you'd be unlikely to notice unless you happened to be looking at the logs of the GitHub Action.

File: cli-files/go-cat/cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+			data, err := os.ReadFile(path)
+			if err != nil {
+				fmt.Println("Can't read file:", path)
```
Comment (illicitonion) [inline cli-files/go-cat/cmd/root.go:29]: We typically write errors to `stderr` rather than `stdout` (which `fmt.Println` writes to be default).

In general, stdout is for "the actual output of my program" and stderr is for "information like progress messages or error descriptions on the way". We have this distinction because otherwise someone may run `go-cat /some/file/which/does/not/exist` and based on the output think that the contents of that file is `Can't read file: /some/file/which/does/not/exist`! This is particularly an issue where you're piping the output of a program to another, e.g. if you ran `cat /does/not/exist | wc -l` to find out how many lines were in the file, and got an answer of `1`, that may be misleading, but the user may never see your error message!

Comment (illicitonion) [inline cli-files/go-cat/cmd/root.go:29]: It's often (but not always!) useful to also include the error you encountered in what you print - knowing whether the error was "file does not exist" or "didn't have permission to read file" may change how you respond to try to fix the issue as a user.

There are times we don't want to do this (e.g. where the error message may contain secret information we don't want to share), but in this case I think it's probably more useful to include :)

File: cli-files/go-cat/cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+				return err
+			}
+			//command that takes a path to a file as an argument
+			os.Stdout.Write(data)
+			out := cmd.OutOrStdout()
+			out.Write(data)
```
Comment (illicitonion) [inline cli-files/go-cat/cmd/root.go:35]: It looks like you're writing the data to stdout twice (that your `os.Stdout.Write` and `cmd.OutOrStdout.Write` are doing the same thing)? Is that intended?

File: cli-files/go-cat/cmd/root.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	return cmd
+}
+
+func Execute() {
+	rootCmd := NewRootCmd()
+	err := rootCmd.Execute()
+	if err != nil {
+		os.Exit(1)
```
Comment (illicitonion) [inline cli-files/go-cat/cmd/root.go:47]: Currently you're only using `err` here as a boolean (was it nil or non-nil) - it may be useful to do your error _printing_ in this function as well.

Right now your `RootCmd` only handles on error, so you only have one print in it, but if you moved the printing of the error into this function, if you introduced more errors to report in your `RootCmd`, you wouldn't have to put your print logic there each time, you could just return an error from `RootCmd.Execute` and any printing you want to do would happen for you because it lived here :)

File: cli-files/go-cat/cmd/root_test.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"testing"
+	"bytes"
+
+	"io/ioutil"
+
+)
+
+func Test_ExecuteCommand(t *testing.T) {
```
Comment (illicitonion) [inline cli-files/go-cat/cmd/root_test.go:11]: I'm not quite sure what this test is trying to test :) We can talk this through on a call soon, but something to think about:

This seems to be testing "If you don't pass a file, we don't print anything", but that's not super clear - a couple of reasons for this are:
1. We're not comparing the output with an empty string, we're giving it a starter value and comparing against that - starting with an empty string, and making sure it stays empty would make it more clear that that's what we're doing.
2. This is testing an error case, but we're not testing that either `Execute` returns an error or that an error message is printed. We should probably check at least one of these things :)

File: cli-files/go-cat/cmd/root_test.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+		t.Fatalf("expected \"%s\" got \"%s\"", "test", string(out))
+	}
+}
+
+func Test_ExecuteCommandWithFile(t *testing.T) {
+	cmd := NewRootCmd()
+	b := bytes.NewBufferString("")
+	cmd.SetOut(b)
+	cmd.SetArgs([]string{"../assets/rain.txt"})
+	cmd.Execute()
```
Comment (illicitonion) [inline cli-files/go-cat/cmd/root_test.go:31]: `cmd.Execute()` may return an error, so we should check whether it does and make sure it's `nil`. If `cmd.Execute()` returned non-`nil` value, we'd want to know about it here - it may help us understand why `b` contained different bytes than we were expecting!

(Some programming languages would _force_ you to do this, and Go does _if a function returns more than one value_, but the case where a function _just_ returns an `error` it lets you ignore it, which is a little annoying!)

File: cli-files/go-cat/cmd/root_test.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+	out, err := ioutil.ReadAll(b)
```
Comment (illicitonion) [inline cli-files/go-cat/cmd/root_test.go:32]: This `ReadAll` totally works, and is a fine thing to do, but it forces us to handle an error that can never actually happen.

The reason `ioutil.ReadAll` returns an error is because it uses the `Read` interface, which may return an error. And the reason _that_ does is because `Read` may involve reading a file (which may fail), talking on the network (which may fail), or other things.

With our `bytes.Buffer`, we _know_ that `Read` can't fail, and so `bytes.Buffer` actually has some convenience methods on it - you can use `b.String()` or `b.Bytes()` to get a `string` or `[]byte` respectively, which means you don't need to use `ioutil.ReadAll(b)` and then check whether it had an error, instead you can jump straight to the type you want with no error check required!

**Review by illicitonion — COMMENTED** [review]
I've only looked at `go-cat` so far, but thought I'd give you some feedback to think about / act on until I can take a look at the rest :)

This is looking really good - you have a program which does what it's meant to, and some tests which test that it works, congrats!

I've left a few comments, and they're all things we haven't really taught you, and that folks writing their first Go CLIs tend to not know about - hopefully the comments I've left are clear enough for you to fix up in your code (and maybe even apply to your other programs!), but of course we can talk through anything you're not sure about :)

Keep up the good work!

File: cli-files/go-ls/cmd/root.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+				dir = args[0]
+			}
+
+			fileInfo, err := os.Stat(dir)
+			if err != nil {
+				return err
+			}
+			if !fileInfo.IsDir() {
+
+				_, file := filepath.Split(dir)
```
Comment (illicitonion) [inline cli-files/go-ls/cmd/root.go:29]: What would happen if you passed a path containing a slash as an argument? Or an absolute path? e.g. `assets/dew.txt` or `/usr/bin/true`? What would you expect to happen? What does `ls` do?

File: cli-files/go-ls/cmd/root.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+				fmt.Fprintln(cmd.OutOrStdout(), file)
+				return nil
+			}
+
+			files, err := os.ReadDir(dir)
+			if err != nil {
+				return err
+			}
+			for _, file := range files {
+				fmt.Fprintln(cmd.OutOrStdout(), file.Name())
```
Comment (illicitonion) [inline cli-files/go-ls/cmd/root.go:39]: Same question about paths containing slashes

File: cli-files/go-ls/cmd/root_test.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	cmd.SetArgs([]string{})
+	cmd.Execute()
+	out, err := ioutil.ReadAll(b)
+	expected := `root.go
+root_test.go
+`
+	if err != nil {
+		t.Fatal(err)
+	}
+	if string(out) != expected {
```
Comment (illicitonion) [inline cli-files/go-ls/cmd/root_test.go:22]: There's a useful library called `require` that I think makes it a little quicker/easier to write tests in Go:
https://pkg.go.dev/github.com/stretchr/testify/require
Rather than having to write out the `if` _and_ the `t.Fatal`, you can just `require.Equal(t, expected, string(out))` and save yourself some time/hassle :)

File: cli-files/go-ls/cmd/root_test.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
+package cmd
```
Comment (illicitonion) [inline cli-files/go-ls/cmd/root_test.go:1]: It can also be useful to write tests for error cases, to make sure that the user-facing result of an error is high quality.

Can you add a test for something like "passed an argument which didn't exist" to show that the output is useful to the user?

File: http-auth/main.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+	"os"
+	"strings"
+
+	"github.com/joho/godotenv"
+	"golang.org/x/time/rate"
+)
+
+var htmlHead = "<!DOCTYPE html><html>"
+
+func goDotEnvVariable(key string) string {
```
Comment (illicitonion) [inline http-auth/main.go:18]: Do you need to load the file every time a request comes in? Would it be better to load it once at the start of the program?

(I don't think either is _better_, they have different trade-offs - when do you think each is a better solution? Why?)

File: http-auth/main.go [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+	}
+	return os.Getenv(key)
+}
+
+func handlerIndex(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+
+	case "GET":
+		w.Header().Set("Content-Type", "text/html") //text/plain sends a string as response
+		keys, ok := r.URL.Query()["foo"]
```
Comment (illicitonion) [inline http-auth/main.go:32]: Can we support _any_ query parameter here, not just foo?

File: http-auth/main.go [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+		if ok {
+			foo := keys[0]
```
Comment (illicitonion) [inline http-auth/main.go:34]: Can we support multiple values here, rather than just one?

File: http-auth/main.go [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+			w.Write([]byte(fmt.Sprintf("%v<em>Hello, world</em><p>Query parameters:<ul><li>foo: %v</li></ul>\n", htmlHead, html.EscapeString(foo))))
+		} else {
+			w.Write([]byte(fmt.Sprintf("%v<em>Hello, world</em>\n", htmlHead)))
```
Comment (illicitonion) [inline http-auth/main.go:37]: This totally works, but here we're doing separate code-paths for "zero query parameters" and "some query parameters"

Sometimes we _do_ need to treat zero or one differently from other cases, but we tend to try to avoid treating them specially if we can, and factor out the common things into common code.

Partially this is so that if we make changes (let's say we wanted to change "Hello, world" to "Hi!"), we only need to change one place, not multiple (it's annoying to have to do things more than once, and we're also more likely to forget!), but also partially this is because it helps readers to distinguish "What's the same" from "What's special"

Can you restructure this so that the "Hello world" code only exist once, and we _add_ to it if there are query parameters?

File: http-auth/main.go [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+		}
+
+	case "POST":
+		w.Header().Set("Content-Type", "text/html")
+		b, err := io.ReadAll(r.Body)
+		if err != nil {
+			log.Fatal(err)
```
Comment (illicitonion) [inline http-auth/main.go:44]: `log.Fatal` is a pretty extreme thing to do in an HTTP handler, as it will crash the entire program and stop it from serving any further requests. In something like an HTTP server, this is not great - one user doing something weird or broken, or one bug in the server, will mean no one can use it any more!

Typically we only call `log.Fatal` in `fn main()` itself, and only when things are super bad.

Here, I'd recommend a `log.Print` so we can see what went wrong, and then rejecting the request (writing an HTTP status code of 500 maybe)

File: http-auth/main.go [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+		}
+		w.WriteHeader(202)
+		w.Write([]byte(fmt.Sprintf("%v%v\n", htmlHead, html.EscapeString(string(b)))))
+	default:
+		w.Write([]byte("Only GET and POST methods are available"))
```
Comment (illicitonion) [inline http-auth/main.go:49]: There's [an HTTP status code for this](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405) - we should probably use it as well as responding with useful text :)

File: http-auth/main_test.go [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+		request, _ := http.NewRequest(http.MethodGet, "/", nil)
+		response := httptest.NewRecorder()
+		handlerIndex(response, request)
+		assertStatus(t, response.Code, http.StatusOK)
+		assertResponseBody(t, response.Body.String(), "<!DOCTYPE html><html><em>Hello, world</em>\n")
+	})
+	t.Run("GET / (WITH QUERY)", func(t *testing.T) {
+		request, _ := http.NewRequest(http.MethodGet, "/", nil)
+		q := request.URL.Query()
+		q.Add("foo", "<strong>bar</strong>")
```
Comment (illicitonion) [inline http-auth/main_test.go:23]: Good job picking the most involved test-case here (one that involves escaping) - that's a really good mindset to have when testing!

File: http-auth/main_test.go [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+	t.Run("GET /404", func(t *testing.T) {
+		request, _ := http.NewRequest(http.MethodGet, "/404", nil)
+		response := httptest.NewRecorder()
+		server := http.NotFoundHandler()
+		server.ServeHTTP(response, request)
+		assertStatus(t, response.Code, http.StatusNotFound)
+	})
+	t.Run("GET /authenticated (OK)", func(t *testing.T) {
+		request, _ := http.NewRequest(http.MethodGet, "/authenticated", nil)
+		request.Header.Add("Authorization", "Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ==")
```
Comment (illicitonion) [inline http-auth/main_test.go:61]: There's an interesting testing philosophy question here - you could either hard-code the base64-encoded string here, or you could write code to compute it (`base64.StdEncoding.EncodeToString([]byte("Aladdin:open sesame"))`).

Hard-coding is useful because it means you can't accidentally make a bug in your code.

Computing is useful because it makes clear to the reader where the value came from, and that it was constructed properly (and easier to change if we want to e.g. change the username in the future).

In this case, I think what you did (hard-coding) is the best answer - we generally want tests to be as easy to read and verify as being bug-free as possible!

It _may_ be useful to also include a comment like // Computed by running `echo -n Aladdin:open sesame | base64` or similar, so that someone knows how to re-generate the string if, say, the format or required username changed.

File: http-auth/main_test.go [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+		response := httptest.NewRecorder()
+		handlerAuth(response, request)
+		assertStatus(t, response.Code, http.StatusOK)
```
Comment (illicitonion) [inline http-auth/main_test.go:64]: It would be useful to assert on the response body too :)

File: http-auth/main_test.go [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+	})
+	t.Run("GET /authenticated (FAIL)", func(t *testing.T) {
+		request, _ := http.NewRequest(http.MethodGet, "/authenticated", nil)
+		request.Header.Add("Authorization", "Basic dXNlcm5hbWU6cGFzc3dvcmQ==")
+		response := httptest.NewRecorder()
+		handlerAuth(response, request)
+		assertStatus(t, response.Code, http.StatusUnauthorized)
+	})
+	t.Run("GET /limited (OK)", func(t *testing.T) {
```
Comment (illicitonion) [inline http-auth/main_test.go:73]: This is a really well-written test, congrats!
Some of the things about it that are really nice:
1. You've extracted the limiter as something you can [dependency inject](https://en.wikipedia.org/wiki/Dependency_injection) - in real life, maybe you want a much higher rate limit, but for your test, you don't want to have to make a million requests to test it out, so you can inject a more limiting rate limiter. It also makes it super clear to read _why_ 30 is a special value, which is nice.
2. The assertion message when things fail is really clear - it has _all_ of the information you may need as someone reading a test log - it's much better than asserting separately on OKs and 429s, because we can see whether we were missing responses, or whether they were wrong, or something else.

File: http-auth/main_test.go [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+		limiter := rate.NewLimiter(100, 30)
+		testHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Write([]byte("[]"))
+		})
+		request, _ := http.NewRequest(http.MethodGet, "/limited", nil)
+		var recorder []int
```
Comment (illicitonion) [inline http-auth/main_test.go:79]: I'm not sure recorder is the best name for this variable - it looks like it stores status codes? Can you think of a better name?

File: http-auth/main_test.go [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+		handler := handlerLimit(limiter, testHandler)
+		for i := 0; i < 32; i++ {
+			rr := httptest.NewRecorder()
+			handler.ServeHTTP(rr, request)
+			recorder = append(recorder, rr.Code)
+		}
+		fmt.Printf("%v", recorder)
+		countOK := 0
```
Comment (illicitonion) [inline http-auth/main_test.go:87]: Rather than making a slice, and then having to go over every entry in the slice to count them, is there a different data structure we could use to store how many 200/429 response codes we saw?

File: http-auth/main_test.go [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+		count429 := 0
+		for _, v := range recorder {
+			if v == 200 {
+				countOK++
+			} else {
+				count429++
```
Comment (illicitonion) [inline http-auth/main_test.go:93]: Do we know this was a 429? What if it was actually a 500? How would we notice?

File: http-auth/main_test.go [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+			}
+		}
+		got1, got2 := countOK, count429
```
Comment (illicitonion) [inline http-auth/main_test.go:96]: I don't think `got1` and `got2` are clearer names than `countOK` and `count429` - I'd probably not do this renaming and either just use the `count*` variables, or maybe rename the `count*` variables to be `gotOK` and `got429`?

File: http-auth/main_test.go [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+		want1, want2 := 30, 2
+		fmt.Println(got1, got2, want1, want2)
```
Comment (illicitonion) [inline http-auth/main_test.go:98]: Let's remove this debug logging - the error message from the assertion failure should be good enough to understand what went wrong if something went wrong.

File: http-auth/main_test.go [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+		if (got1 != want1) || (got2 != want2) {
+			t.Errorf("response body is wrong, got number of successful requests %v and rejected %v want %v and %v", got1, got2, want1, want2)
```
Comment (illicitonion) [inline http-auth/main_test.go:100]: "response body" isn't quite what's wrong here? Maybe reword the message a little?

File: http-auth/main_test.go [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+		}
+	})
+}
+
+func assertStatus(t testing.TB, got, want int) {
```
Comment (illicitonion) [inline http-auth/main_test.go:105]: Extracting these helper functions is really useful, it helps to make the tests read better, and avoid duplication of code!

**Review by illicitonion — COMMENTED** [review]
Ok, I've reviewed this all now - in general, it's looking great, good job! In particular, you've been writing really high-quality tests, which isn't always easy - congrats!

Where a suggestion applies to multiple places, I've only left it in once - an exercise for you is to work out where else the comments may apply, and fix things up everywhere.

In general this is looking really good, and some of my comments are more "let's think about the philosophy here for the future" rather than "there's something to improve here now" :)

Good job! 🎉

File: http-auth/main_test.go (hunk 24 above)
Comment (RitaGlushkova) [reply http-auth/main_test.go:93]: The task was to write 503 server is unavailable. I wasn't sure why...

File: http-auth/main_test.go (hunk 23 above)
Comment (RitaGlushkova) [reply http-auth/main_test.go:87]: responsesCodes?

File: http-auth/main_test.go (hunk 22 above)
Comment (RitaGlushkova) [reply http-auth/main_test.go:79]: responsesCodes?

File: http-auth/main_test.go (hunk 24 above)
Comment (RitaGlushkova) [reply http-auth/main_test.go:93]: I could do another if statement instead of else?

File: http-auth/main_test.go (hunk 21 above)
Comment (RitaGlushkova) [reply http-auth/main_test.go:73]: Thank you!!

File: http-auth/main_test.go (hunk 18 above)
Comment (RitaGlushkova) [reply http-auth/main_test.go:23]: do you mean using "foo" "bar"?

File: http-auth/main.go (hunk 17 above)
Comment (RitaGlushkova) [reply http-auth/main.go:49]: I am not sure I understand, do you mean in the default section?

File: http-auth/main.go (hunk 16 above)
Comment (RitaGlushkova) [reply http-auth/main.go:44]: It is a very good point, I did not realise that. Thank you!

Comment (RitaGlushkova) [reply http-auth/main.go:44]: I changed it, but I realised that I can not figure out how to test it because everything I pass into a post request turns into a string and return status 202. So I actually can not see what is happening if it fails.

File: http-auth/main.go (hunk 15 above)
Comment (RitaGlushkova) [reply http-auth/main.go:37]: `	case "GET":
		w.Header().Set("Content-Type", "text/html") //text/plain sends a string as response
		keys, ok := r.URL.Query()["foo"]
		defaultGreeting := "<em>Hello, world</em>"
		if ok {
			foo := keys[0]
			w.Write([]byte(fmt.Sprintf("%v%v<p>Query parameters:<ul><li>foo: %v</li></ul>\n", htmlHead, defaultGreeting, html.EscapeString(foo))))
		} else {
			w.Write([]byte(fmt.Sprintf("%v%v\n", htmlHead, defaultGreeting)))
		}`

File: cli-files/go-ls/cmd/root.go (hunk 8 above)
Comment (RitaGlushkova) [reply cli-files/go-ls/cmd/root.go:29]: Changed to make ls and go-ls to print out directory

File: cli-files/go-ls/cmd/root.go (hunk 9 above)
Comment (RitaGlushkova) [reply cli-files/go-ls/cmd/root.go:39]: Can we discuss this point? I am trying both ls and go-ls they behave the same way.

File: cli-files/go-cat/cmd/root.go (hunk 2 above)
Comment (RitaGlushkova) [reply cli-files/go-cat/cmd/root.go:29]: Can I just return `err` here and then print `err` in `func main()`?

File: cli-files/go-cat/cmd/root.go (hunk 1 above)
Comment (illicitonion) [reply cli-files/go-cat/cmd/root.go:26]: Looks good! (Alternatively, the real `cat` prints out _all_ of the files it's passed, one after the other, which is also reasonable :))

File: cli-files/go-cat/cmd/root.go (hunk 2 above)
Comment (illicitonion) [reply cli-files/go-cat/cmd/root.go:29]: Yep!

File: cli-files/go-ls/cmd/root.go (hunk 9 above)
Comment (illicitonion) [reply cli-files/go-ls/cmd/root.go:39]: You're right, this was a bad comment! :D Apologies!

File: http-auth/main.go (hunk 16 above)
Comment (illicitonion) [reply http-auth/main.go:44]: Testing weird error cases is an interesting problem!

Think about when this error triggers? When reading from `Body` returns an error.

In this case, we're constructing the request (our call to `http.NewRequest`, and we pass in the `Body` when constructing it, so we could pass in an `io.Reader` which returns an error when you read from it?

File: http-auth/main.go (hunk 17 above)
Comment (illicitonion) [reply http-auth/main.go:49]: Yeah, in the default section - if the method wasn't `GET` or `POST`

File: http-auth/main_test.go (hunk 18 above)
Comment (illicitonion) [reply http-auth/main_test.go:23]: I mean that `<strong>bar</strong>` was a good example of a querystring to use - you could have just used `bar`, but that would have tested less of your code than also including some HTML that needs escaping :)

File: http-auth/main_test.go (hunk 22 above)
Comment (illicitonion) [reply http-auth/main_test.go:79]: Sounds great!

File: cli-files/go-ls/cmd/root.go (hunk 9 above)
Comment (RitaGlushkova) [reply cli-files/go-ls/cmd/root.go:39]: but I just realised that ls and go-ls behave differently on `go-ls dew.txt for_you.txt ` and `ls dew.txt for_you.txt`. First prints only first file name, ls prints both...

File: http-auth/main_test.go (hunk 24 above)
Comment (RitaGlushkova) [reply http-auth/main_test.go:93]:
```
if v == 429 {
				got429++
			} else {
				t.Errorf("did not get correct status code. Got %v, want 200 or 429.", v)
			}
```

File: cli-files/go-ls/cmd/root.go (hunk 9 above)
Comment (RitaGlushkova) [reply cli-files/go-ls/cmd/root.go:39]: resolved

File: http-auth/main.go (hunk 12 above)
Comment (RitaGlushkova) [reply http-auth/main.go:18]: Do you mean to make these variables global to the whole program at the beginning? and then use those in http request when needed. 
I guess it is better when we require Authorization in multiple endpoints. Would like to discuss it in more details.
var username = goDotEnvVariable("AUTH_USERNAME")
var password = goDotEnvVariable("AUTH_PASSWORD")

File: http-auth/main_test.go (hunk 24 above)
Comment (RitaGlushkova) [reply http-auth/main_test.go]: Honestly stuck here...

Comment (RitaGlushkova) [reply http-auth/main_test.go]: if v != 200 && v != 429 {
				t.Errorf("did not get correct of responses code. Got code: %v wanted 200 or 429", v)
			}

File: http-auth/main_test.go (hunk 23 above)
Comment (RitaGlushkova) [reply http-auth/main_test.go]: Honestly, can not think of any...

Comment (illicitonion) [reply http-auth/main_test.go]: I would probably use a `map[int]int` where the keys are status codes and the values are counts :) Then you can both get `counts[200]` and see if there are any keys which aren't `200` or `429`

Comment (RitaGlushkova) [reply http-auth/main_test.go]: Yes of course! 
I haven't used map in GO much, so doesn't come to my mind naturally just yet...

File: cli-files/go-cat/cmd/root.go [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+	"io"
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+func CatFiles(out io.Writer, args []string) error {
+	for _, arg := range args {
+
+		data, err := os.ReadFile(arg)
```
Comment (illicitonion) [inline cli-files/go-cat/cmd/root.go:17]: Using `os.ReadFile` forces you to allocate memory for the _whole_ file - imagine you're running `cat` on a file which is 100GB large - to print it out you don't actually need to read the whole 100GB at once, but your program needs to, which means your computer would need 100GB of RAM to spare.

Can you think of a way to read the file from disk, and write it to stdout, which _doesn't_ require you to read the whole thing into one slice?

**Review by illicitonion — COMMENTED** [review]
This is looking really good, nicely done! I left a few comments, mostly around some practices around testing :)

File: http-auth/main.go [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+		}
+
+	case "POST":
+		w.Header().Set("Content-Type", "text/html")
+		b, err := io.ReadAll(r.Body)
+		if err != nil {
+			log.Print(err)
+			w.WriteHeader(500)
+		}
+		w.WriteHeader(202)
```
[the remark this answers is not in the document]
Comment (RitaGlushkova) [reply http-auth/main.go:60]:
```
if err != nil {
			log.Print("Error reading body of POST to index",err)
			w.WriteHeader(500)
			return
		}
```
